### PR TITLE
Add nested fragments

### PR DESCRIPTION
**Pull request by elidhu**
I have added nested fragment parsing. I have not had to touch any tests except:

- Removing the "double open" test as it is no longer an error.
- Adding a "halt while open" test that I noticed was missing.
- Adding a "nested fragment" test.

The original behaviour should be identical as far as the tests go.

The new behaviour can be seen in the example I added. The most notable part is the handling of nested fragment tags when outputting a fragment that has children. Nested tags are excluded as if they weren't there.

Let me know your thoughts.

Thanks!

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -155,118 +155,106 @@ pub fn extract_fragments(
...
+            break;
+        } else if let Some(fragment) = fragment_stack.last_mut() {
+            fragment.body.push_str(content);
+            fragment.body.push('\n');
         }
     }
 
+    if !fragment_stack.is_empty() {
+        return Err(FileError {
+            err_type: ParseError::DoubleOpen,
```
Comment (nickpascucci) [inline src/lib.rs:250]: **request:** Can you rename this error variant to reflect the changed semantics? Perhaps something like `UnclosedFragment`.

File: src/lib.rs [hunk 2]
```diff
@@ -742,6 +730,60 @@ def main():
...
+    # >@
+    End of outer
+    # >@",
+            "test.py",
+            &SymbolKey::default(),
+        );
+
+        let fragments = fragments.expect("Expected no parse errors");
+        assert!(
+            fragments.len() == 3,
```
Comment (nickpascucci) [inline src/lib.rs:753]:
```suggestion
            fragments.len() == 3, // Two fragments in addition to the full file
```

File: src/lib.rs [hunk 3]
```diff
@@ -742,6 +730,60 @@ def main():
...
+            fragments[1].id == *"qux",
+            "Unexpected ID {:?}",
+            fragments[1].id
+        );
+        assert_eq!(
+            fragments[2].body,
+            "    Start of outer
+        Start of inner
+        End of inner
+    End of outer",
```
Comment (nickpascucci) [inline src/lib.rs:778]:
```suggestion
    End of outer",
            "Expected nested fragment markers to be removed"
```

**Review by nickpascucci — CHANGES_REQUESTED** [review]
Thanks for your PR! This looks good, nice work.

Before I approve, can I ask you to update the README with:

- A mention that fragments can be nested
- Removing the "future work" item
- Adding yourself to the "Contributors" section

File: src/lib.rs (hunk 1 above)
Comment (elidhu) [reply src/lib.rs]: Great pickup!

**Comment by elidhu** [comment]
Addressed all of the above, and updated the README.

Thanks for taking the time to review!

**Comment by elidhu** [comment]
I did have the thought while I was making the changes that this adds support for "nested" fragments, not necessarily "overlapping". This would required some sort of ID on the closing tag (I guess).

I'm not sure if that was an important distinction

**Comment by nickpascucci** [comment]
Yes, good point — I don't think it's a hugely important distinction, and as you said, once the fragment structure is no longer tree-like it becomes necessary to annotate the closing tags. This seems like a good next step, adding a nice feature without complecting the system too much.

**Review by nickpascucci — APPROVED** [review]

**Comment by elidhu** [comment]
RIght, had to fix all my signatures. It's good-to-go now!

**Comment by elidhu** [comment]
Would you be able to get this merged and cut a release? 🙏

**Comment by nickpascucci** [comment]
@elidhu sorry for the wait — I've just published v0.3.0 which includes your change 🎉

**Pull request merged**